drv/transceivers-server/src/udp.rs

```diff
@@ -217,397 +215,572 @@ impl ServerImpl {
+    /// This should be called as the last operation before sending the response
+    /// since error details is the last of the trailing data. This function
+    /// panics if `None` is the supplied `failure_type` and the `result`
+    /// contains any bits set in the failure mask.
+    fn handle_errors(
         &mut self,
         modules: ModuleId,
+        result: ModuleResult,
+        failure_type: Option<HwError>,
+        data_idx: usize,
         out: &mut [u8],
-    ) -> Result<usize, Error> {
-        let fpga = get_fpga(modules)?;
-        let fpga = self.transceivers.fpga(fpga);
-
-        // This is a bit awkward: the FPGA will get _every_ module's
-        // status (for the given FPGA), then we'll unpack to only the
-        // ones that we care about
-        let enable: U16<LittleEndian> =
-            fpga.read(Addr::QSFP_POWER_EN0).map_err(|_e| {
-                Error::StatusFailed(HwError::PowerEnableReadFailed)
-            })?;
-        let resetl: U16<LittleEndian> = fpga
-            .read(Addr::QSFP_MOD_RESETL0)
-            .map_err(|_e| Error::StatusFailed(HwError::ResetLReadFailed))?;
-        let lpmode: U16<LittleEndian> = fpga
-            .read(Addr::QSFP_MOD_LPMODE0)
-            .map_err(|_e| Error::StatusFailed(HwError::LpModeReadFailed))?;
-        let presentl: U16<LittleEndian> = fpga
-            .read(Addr::QSFP_MOD_MODPRSL0)
-            .map_err(|_e| Error::StatusFailed(HwError::ModPrsLReadFailed))?;
-        let irql: U16<LittleEndian> = fpga
-            .read(Addr::QSFP_MOD_INTL0)
-            .map_err(|_e| Error::StatusFailed(HwError::IntLReadFailed))?;
-        let pg: U16<LittleEndian> = fpga
-            .read(Addr::QSFP_POWER_GOOD0)
-            .map_err(|_e| Error::StatusFailed(HwError::PgReadFailed))?;
-        let pg_to: U16<LittleEndian> = fpga
-            .read(Addr::QSFP_POWER_GOOD_TIMEOUT0)
-            .map_err(|_e| Error::StatusFailed(HwError::PgTimeoutReadFailed))?;
-        let pg_lost: U16<LittleEndian> = fpga
-            .read(Addr::QSFP_POWER_GOOD_LOST0)
-            .map_err(|_e| Error::StatusFailed(HwError::PgLostReadFailed))?;
-
-        // Write one bitfield per active port in the ModuleId
+    ) -> (usize, ModuleId) {
+        let mut error_idx: usize = data_idx;
+        // any modules at index 32->63 are not currently supported.
+        let invalid_modules = ModuleId(0xffffffff00000000);
+        let requested_invalid_modules = ModuleId(modules.0 & invalid_modules.0);
+        // iterate through the results and populate errors in the response buffer
+        // failure: whatever `HwError` specified by `failure_type`
+        // error: fpga communication issue
+        for module in modules.to_indices().map(LogicalPort) {
+            if module <= LogicalPortMask::MAX_PORT_INDEX {
+                if result.failure().is_set(module) {
+                    match failure_type {
+                        Some(e) => {
+                            out[error_idx] = e as u8;
```

Comment:
I'm not sure casting with `as` is quite right. I know support for that [was added](https://doc.rust-lang.org/reference/items/enumerations.html#casting), but I think we probably want to lean on `hubpack::serialize` here too. That would allow us to add variants that carry data, without changing this code.

Not a huge deal I guess, since this cast should fail if that becomes the case.

Reply:
Done! I changed all the error serializations in this function to be of the form
```rust
let err_size = hubpack::serialize(&mut out[error_idx..], &some_err).unwrap();
error_idx += err_size;
```
which should help make this a little more future proof.